### PR TITLE
feat: asset info

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

Wallets need to display asset info (name, logo, ticker etc.).

This branch is based on #140 branch.

# Proposed Solution

Add `Wallet.assets$` that emits asset info for all assets found in `Wallet.balance.total$`

### CODE REVIEW

File: packages/blockfrost/src/blockfrostWalletProvider.ts [hunk 1]
```diff
@@ -8,93 +7,36 @@ import {
...
+  });
 
 /**
  * Connect to the [Blockfrost service](https://docs.blockfrost.io/)
  *
  * @param {Options} options BlockFrostAPI options
  * @returns {WalletProvider} WalletProvider
+ * @throws {ProviderFailure}
  */
 export const blockfrostProvider = (options: Options, logger = dummyLogger): WalletProvider => {
```
Comment (rhyslbw) [inline packages/blockfrost/src/blockfrostWalletProvider.ts:37]:
```suggestion
export const blockfrostWalletProvider = (options: Options, logger = dummyLogger): WalletProvider => {
```

File: packages/blockfrost/src/util.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  throw new ProviderError(ProviderFailure.Unknown, error, `status_code: ${status_code}`);
+};
+
+export const withProviderErrors = <T>(providerImplementation: T) =>
+  Object.keys(providerImplementation).reduce((provider, key) => {
+    provider[key] = (...args: any[]) => (providerImplementation as any)[key](...args).catch(toProviderError);
+    return provider;
+  }, {} as any) as T;
+
+// TODO: test this
```
Comment (rhyslbw) [inline packages/blockfrost/src/util.ts:55]:
```suggestion
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/blockfrost/src/blockfrostWalletProvider.ts [hunk 3]
```diff
@@ -306,12 +248,38 @@ export const blockfrostProvider = (options: Options, logger = dummyLogger): Wall
     ];
   };
 
+  const fetchJsonMetadata = async (txHash: Cardano.Hash16): Promise<Cardano.MetadatumMap | null> => {
+    try {
+      const response = await blockfrost.txsMetadata(txHash);
+      return response.reduce((map, metadatum) => {
+        // Not sure if types are correct, missing 'label', but it's present in docs
```
Comment (rhyslbw) [inline packages/blockfrost/src/blockfrostWalletProvider.ts:255]: It was fixed in https://github.com/blockfrost/blockfrost-js/releases/tag/v2.0.1. [2.0.0]( https://github.com/blockfrost/blockfrost-js/releases/tag/v2.0.0) appears to only have one breaking change, so the upgrade should be routine.

File: packages/blockfrost/src/blockfrostWalletProvider.ts [hunk 4]
```diff
@@ -306,12 +248,38 @@ export const blockfrostProvider = (options: Options, logger = dummyLogger): Wall
...
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const { json_metadata, label } = metadatum as any;
+        if (!json_metadata || !label) return map;
+        // not sure how Blockfrost represents integer metadatum. Might need to convert to bigint here.
```
Comment (rhyslbw) [inline packages/blockfrost/src/blockfrostWalletProvider.ts:259]: Are we missing a test to cover this? We can dig up an example tx on _testnet_ that includes a large integer to use as a mock.

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/blockfrost/src/blockfrostWalletProvider.ts (hunk 4 above)
Comment (mkazlauskas) [reply packages/blockfrost/src/blockfrostWalletProvider.ts:259]: Yep, we are missing a test for this. That would be nice, how does one dig up tx'es?

**Review by rhyslbw — APPROVED** [review]
Great work @mkazlauskas

**Pull request merged**